libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,907 @@
+///
+/// Thus the maximum number of pages in one more than this number.
+const MAX_PAGE: usize = 63;
+
+/// Maximum number of keys.
+const MAX_KEY: usize = 4095;
+
+/// Maximum length in bytes of a user payload.
+const MAX_VALUE_LEN: usize = 1023;
+
+/// Maximum number of updates per transaction.
+const MAX_UPDATES: usize = 31;
+
+/// Maximum number of words per virtual page.
+const MAX_VIRT_PAGE_SIZE: usize = div_ceil(MAX_PAGE_SIZE, WORD_SIZE) - CONTENT_WORD;
+
+/// Word with all bits set to one.
+const ERASED_WORD: u32 = 0xffffffff;
+
+/// Helpers for a given storage configuration.
+#[derive(Clone, Debug)]
+pub struct Format {
+    /// The size in bytes of a page in the storage.
+    page_size: usize,
+
+    /// The number of pages in the storage.
+    num_pages: usize,
+
+    /// The maximum number of times a page can be erased.
+    max_page_erases: usize,
+}
+
+impl Format {
+    /// Extracts the format from a storage.
+    ///
+    /// Returns `None` if the storage is not [supported].
+    ///
+    /// [supported]: struct.Format.html#method.is_storage_supported
+    pub fn new<S: Storage>(storage: &S) -> Option<Format> {
+        if Format::is_storage_supported(storage) {
+            Some(Format {
+                page_size: storage.page_size(),
+                num_pages: storage.num_pages(),
+                max_page_erases: storage.max_page_erases(),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Returns whether a storage is supported.
+    ///
+    /// A storage is supported if the following conditions hold:
+    /// - The size of a word is 4 bytes.
+    /// - The size of a word evenly divides the size of a page.
+    /// - A page contains at least 8 words.
+    /// - A page contains at most [`MAX_PAGE_SIZE`] bytes.
+    /// - There is at least 3 pages.
+    /// - There is at most [`MAX_PAGE`]` + 1` pages.
+    /// - A word can be written at least twice between erase cycles.
+    /// - A page can be erased at most [`MAX_CYCLE`] times.
+    ///
+    /// [`MAX_PAGE_SIZE`]: constant.MAX_PAGE_SIZE.html
+    /// [`MAX_PAGE`]: constant.MAX_PAGE.html
+    /// [`MAX_CYCLE`]: constant.MAX_CYCLE.html
+    fn is_storage_supported<S: Storage>(storage: &S) -> bool {
+        let word_size = storage.word_size();
+        let page_size = storage.page_size();
+        let num_pages = storage.num_pages();
+        let max_word_writes = storage.max_word_writes();
+        let max_page_erases = storage.max_page_erases();
+        word_size == 4
+            && page_size % word_size == 0
+            && (8 * word_size <= page_size && page_size <= MAX_PAGE_SIZE)
+            && (3 <= num_pages && num_pages <= MAX_PAGE + 1)
+            && max_word_writes >= 2
+            && max_page_erases <= MAX_CYCLE
+    }
+
+    /// The size of a word in bytes.
+    pub fn word_size(&self) -> usize {
+        WORD_SIZE
+    }
+
+    /// The size of a page in bytes.
+    ///
+    /// We have `32 <= self.page_size() <= MAX_PAGE_SIZE`.
```

Comment:
I see the check `8 * word_size <= page_size` in the `is_storage_supported` function. I guess we don't plan for anything different than `word_size == 32`, but would be worth clarifying in case the word size is something else.

Reply:
You probably meant `word_size == 4`, in which case yes, we don't plan to support other architectures for now. But all the code is ready to support them. I clarified in the few places where units are bytes instead of words that the numbers are valid only if words are 4 bytes.

Reply:
If we only plan to support that, I don't see the point of having a `self.word_size()` function, as opposed to directly using the `WORD_SIZE` constant everywhere.

Reply:
I'm resolving in favor of duplicate question already answered here: https://github.com/google/OpenSK/pull/172#discussion_r502290026